### PR TITLE
🎉 ApprenticeCommitment Outer API

### PR DESCRIPTION
**Pull request by benagain**
Project structure for the Apprentice Commitment API.

https://skillsfundingagency.atlassian.net/wiki/spaces/NDL/pages/2440101994/eCommitments+-+C4+Component+ViewPoint

### CODE REVIEW

**Review by MarieDale3 — DISMISSED** [review]

File: src/SFA.DAS.ApprenticeCommitments/InnerApi/Requests/CreateApprenticeshipRequest.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    public class CreateApprenticeshipRequest : IPostApiRequest
+    {
+        public string PostUrl => "/apprenticeships";
+
+        public object Data { get; set; }
+    }
+
+    public class CreateApprenticeshipRequestData
+    {
+        public Guid RequestId { get; set; }
```
Comment (Pauljgraham) [inline src/SFA.DAS.ApprenticeCommitments/InnerApi/Requests/CreateApprenticeshipRequest.cs:15]: This is called RegistrationId in the inner API, not sure which name is right

File: src/SFA.DAS.ApprenticeCommitments/Infrastructure/ApprenticeCommitmentsHealthCheck.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+using System.Threading.Tasks;
+
+namespace SFA.DAS.ApprenticeCommitments.Infrastructure
+{
+    public class ApprenticeCommitmentsHealthCheck : IHealthCheck
+    {
+        private const string HealthCheckResultDescription = "Employer Incentives Api Health Check";
+        private readonly ApprenticeCommitmentsService service;
+
+        public ApprenticeCommitmentsHealthCheck(ApprenticeCommitmentsService service)
```
Comment (Pauljgraham) [inline src/SFA.DAS.ApprenticeCommitments/Infrastructure/ApprenticeCommitmentsHealthCheck.cs:16]: Should we not be injecting the Interface into the service

File: src/SFA.DAS.ApprenticeCommitments/Application/Services/ApprenticeCommitmentsService.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            }
+        }
+
+        public async Task CreateApprenticeship(Guid guid, long apprenticeshipId, string email)
+        {
+            var response = await _client.Post<CreateApprenticeshipResponse>(new CreateApprenticeshipRequest
+            {
+                Data = new CreateApprenticeshipRequestData
+                {
+                    RequestId = guid,
```
Comment (Pauljgraham) [inline src/SFA.DAS.ApprenticeCommitments/Application/Services/ApprenticeCommitmentsService.cs:36]: should it be registrationId ?

File: src/SFA.DAS.ApprenticeCommitments/Application/Services/ApprenticeCommitmentsService.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+using SFA.DAS.ApprenticeCommitments.Configuration;
+using SFA.DAS.ApprenticeCommitments.InnerApi.Requests;
+using SFA.DAS.SharedOuterApi.Interfaces;
+using System;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.ApprenticeCommitments.Application.Services
+{
+    public class ApprenticeCommitmentsService //: IEmployerIncentivesService
```
Comment (Pauljgraham) [inline src/SFA.DAS.ApprenticeCommitments/Application/Services/ApprenticeCommitmentsService.cs:10]: Do we not need the interface here

File: src/SFA.DAS.ApprenticeCommitments.Api/nlog.config [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
+      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+      xmlns:xdt="http://schemas.microsoft.com/XML-Document-Transform">
+    <extensions>
+        <add assembly="SFA.DAS.NLog.Targets.Redis" />
+    </extensions>
+    <variable name="appName" value="das-employer-incentives-outer-api"/>
```
Comment (Pauljgraham) [inline src/SFA.DAS.ApprenticeCommitments.Api/nlog.config:8]: this should be 'das-apprentice-commitments-outer-api'

File: src/SFA.DAS.ApprenticeCommitments.Api/nlog.Debug.config [hunk 6]
```diff
@@ -0,0 +1,19 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
+<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
+      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+      xmlns:xdt="http://schemas.microsoft.com/XML-Document-Transform">
+  <extensions>
+    <add assembly="SFA.DAS.NLog.Targets.Redis" />
+  </extensions>
+  <variable name="appName" value="das-employer-incentives-outer-api"/>
```
Comment (Pauljgraham) [inline src/SFA.DAS.ApprenticeCommitments.Api/nlog.Debug.config:8]: this should be 'das-apprentice-commitments-outer-api'

File: src/SFA.DAS.ApprenticeCommitments.Api/SFA.DAS.ApprenticeCommitments.Api.csproj [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        <Content Update="nlog.config">
+            <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+        </Content>
+        <Content Update="nlog.config">
+          <ExcludeFromSingleFile>true</ExcludeFromSingleFile>
+          <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+          <CopyToPublishDirectory>PreserveNewest</CopyToPublishDirectory>
+          <TransformOnBuild>true</TransformOnBuild>
+        </Content>
+    </ItemGroup>
```
Comment (Pauljgraham) [inline src/SFA.DAS.ApprenticeCommitments.Api/SFA.DAS.ApprenticeCommitments.Api.csproj:55]: curious why we have 2 content sections for nlog.config. I don't really recognise the second section (well I've not seen that type of block before)

File: src/SFA.DAS.ApprenticeCommitments.Api/SFA.DAS.ApprenticeCommitments.Api.csproj [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    <ItemGroup>
+      <None Include="nlog.Debug.config">
+        <IsTransformFile>true</IsTransformFile>
+        <DependentUpon>nlog.config</DependentUpon>
+      </None>
+      <None Include="nlog.Release.config">
+        <IsTransformFile>true</IsTransformFile>
+        <DependentUpon>nlog.config</DependentUpon>
+      </None>
+    </ItemGroup>
```
Comment (Pauljgraham) [inline src/SFA.DAS.ApprenticeCommitments.Api/SFA.DAS.ApprenticeCommitments.Api.csproj:21]: this is new to me, I didn't think we did this. I haven't spotted us doing this before to the nlog files

File: src/SFA.DAS.ApprenticeCommitments.Api/AppStart/AddServiceRegistrationExtension.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+{
+    public static class AddServiceRegistrationExtension
+    {
+        public static void AddServiceRegistration(this IServiceCollection services)
+        {
+            services.AddHttpClient();
+            services.AddTransient<IAzureClientCredentialHelper, AzureClientCredentialHelper>();
+
+            services.AddTransient(typeof(IInternalApiClient<>), typeof(InternalApiClient<>));
+            services.AddTransient<ApprenticeCommitmentsService>();
```
Comment (Pauljgraham) [inline src/SFA.DAS.ApprenticeCommitments.Api/AppStart/AddServiceRegistrationExtension.cs:18]: Do we need to wire this up with a supporting Interface

**Review by Pauljgraham — COMMENTED** [review]
Added some discussion points.

File: src/SFA.DAS.ApprenticeCommitments/InnerApi/Requests/CreateApprenticeshipRequest.cs (hunk 1 above)
Comment (benagain) [reply src/SFA.DAS.ApprenticeCommitments/InnerApi/Requests/CreateApprenticeshipRequest.cs:15]: I *thought* I had copied that from the inner API, obviously not 😆

File: src/SFA.DAS.ApprenticeCommitments.Api/SFA.DAS.ApprenticeCommitments.Api.csproj (hunk 7 above)
Comment (benagain) [reply src/SFA.DAS.ApprenticeCommitments.Api/SFA.DAS.ApprenticeCommitments.Api.csproj:55]: That's how it was in the `EmployerIncentives` project that was used as inspiration.  Looking at it, seems to be unnecessary - the second section will overwrite the first section.

File: src/SFA.DAS.ApprenticeCommitments.Api/SFA.DAS.ApprenticeCommitments.Api.csproj (hunk 8 above)
Comment (benagain) [reply src/SFA.DAS.ApprenticeCommitments.Api/SFA.DAS.ApprenticeCommitments.Api.csproj:21]: I haven't encountered using overide files for NLog either, but the other APIM projects do use them.

File: src/SFA.DAS.ApprenticeCommitments/Application/Services/ApprenticeCommitmentsService.cs (hunk 4 above)
Comment (benagain) [reply src/SFA.DAS.ApprenticeCommitments/Application/Services/ApprenticeCommitmentsService.cs:10]: Once there is a need for the interface then yes.  As of the state of the project right now there is no need for one, is there?

File: src/SFA.DAS.ApprenticeCommitments.Api/nlog.config (hunk 5 above)
Comment (benagain) [reply src/SFA.DAS.ApprenticeCommitments.Api/nlog.config:8]: Updated

File: src/SFA.DAS.ApprenticeCommitments/Infrastructure/ApprenticeCommitmentsHealthCheck.cs (hunk 2 above)
Comment (benagain) [reply src/SFA.DAS.ApprenticeCommitments/Infrastructure/ApprenticeCommitmentsHealthCheck.cs:16]: See https://github.com/SkillsFundingAgency/das-apim-endpoints/pull/232#discussion_r556401338

File: src/SFA.DAS.ApprenticeCommitments.Api/AppStart/AddServiceRegistrationExtension.cs (hunk 9 above)
Comment (benagain) [reply src/SFA.DAS.ApprenticeCommitments.Api/AppStart/AddServiceRegistrationExtension.cs:18]: See https://github.com/SkillsFundingAgency/das-apim-endpoints/pull/232#discussion_r556401338

**Review by Pauljgraham — DISMISSED** [review]

**Pull request merged**